fontdir: Consider scalable fonts in index fonts.dir

EmilioPeJu opened the pull request:
mkfontdir requires the file fonts.scale to consider scalable fonts, thus, mkfontscale should be run before.

This change affects to everyone using the option enableFontDir

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
- The need to use scalable fonts in an old application (using the X11 core font API instead of font-config)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

EmilioPeJu committed:
nixos/fontdir: gather more font formats

- Fix wrong order in which font indexes are created
mkfontdir requires the file fonts.scale to consider scalable fonts,
thus, mkfontscale should be run before

- Search more font formats, in particular, bit-mapped formats

rnhmjoj commented:
> mkfontdir requires the file fonts.scale to consider scalable fonts, thus, mkfontscale should be run before.

Ah, I've always thought `fonts.dir` only included bitmap fonts and `fonts.scale` vector fonts, but mkfontdir(1) is pretty clear about this: thanks. I wonder if we whould change the order everywhere, there are tons of packages calling the two commands in the wrong order.

In any case, I'm not familiar with this modules: IIUC it should glue all fonts.{scale,dir,alias} together.
I don't understand some of the steps it's doing:

1. collects all `fonts.dir`, ttf and otf fonts to /share/X11-fonts (why not fonts.scale as well?)
2. removes all `fonts.{dir,scale,alias}
3. runs mkfont{dir,scale} in /share/X11-fonts (won't this only detect scalable fonts? the `find` command is not linking pfa, pfb, pcf, bdf, etc..)
4. concatenates all `fonts.alias` files and saves to /share/X11-fonts (makes sense)

hir12111 commented:
Hi @rnhmjoj and thanks for your comment,

1. and 2. You are right, this part could be improved, we don't need to link fonts.dir if we are going to delete it in the next line
The reason why we don't use the `fonts.dir` and `fonts.scale` provide by packages is because they might not provide it, and thus it is safer to recreate them after we collect them all in the same directory
3. You are also right about this, I'm not sure if avoiding bit-mapped fonts is intentional, I think that it would make more sense to collect all the fonts specified by the user (even if they are bit-mapped)

 I could add one more commit to fix those other issues, I'll start testing it

rnhmjoj commented:
> I could add one more commit to fix those other issues, I'll start testing it

That would be great, thanks. I also wanted to ask: what are you using to test the change? I'd like to try it myself.

hir12111 commented:
> > I could add one more commit to fix those other issues, I'll start testing it
> 
> That would be great, thanks. I also wanted to ask: what are you using to test the change? I'd like to try it myself.

The simplest way I can think of,
1. Clone nixpkgs or copy the folder associated to the nixos channel you are using (in my case `/nix/var/nix/profiles/per-user/root/channels/nixos`), let's say the destination path is `/home/user/nixos`
2. Make the changes in that folder
2. Run: `nixos-rebuild -I nixos=/home/user/nixos -I nixpkgs=/home/user/nixos switch`

I assume there are better ways, but I don't know them yet xD

rnhmjoj commented:
Sorry, I meant which program you were using to display the fonts. I only know of xfontsel, which uses the X11 font API.

hir12111 commented:
I'm using the same one for display, I also use `xlsfonts` to confirm the font was taken

hir12111 commented:
I added the most common bit-mapped font formats (which very often are provided compressed), however I was tempted to just search for all regular files, Do you think that could have some bad side effect?

hir12111 commented:
By the way, you probably need to add the X11-fonts package folder to font-path by using `xset +fp /path/to/x11-fonts-output/share/X11-fonts.`, for some reason it's not allowed to do so with `/run/current-system/sw/share/X11-fonts`, I suspect it doesn't like that `fonts.dir` is a link

rnhmjoj commented on nixos/modules/config/fonts/fontdir.nix:
```diff
@@ -5,14 +5,16 @@ with lib;
 let
 
   x11Fonts = pkgs.runCommand "X11-fonts" { preferLocalBuild = true; } ''
+    font_formats=('ttf' 'otf' 'pcf' 'pfa' 'pfb' 'pcf' 'bdf')
```
You wrote `pcf` two times.

hir12111 replied:
ups, sorry about that, I amended the commit

rnhmjoj commented on nixos/modules/config/fonts/fontdir.nix:
```diff
@@ -5,14 +5,16 @@ with lib;
-    find ${toString config.fonts.fonts} \
-      \( -name fonts.dir -o -name '*.ttf' -o -name '*.otf' \) \
-      -exec ln -sf -t "$out/share/X11-fonts" '{}' \;
+    for fmt in "''${font_formats[@]}"; do
```
I think you should be able to use `-regex` to avoid calling `find` multiple times and the `-o`.

hir12111 replied:
I implemented that at the cost of a more complicated code, I'm open to suggestions for a simpler regexp-based way

hir12111 replied:
I amended the commit to simplify a bit more

EmilioPeJu committed:
nixos/fontdir: use regexp to find font files

rnhmjoj committed:
nixos/xserver: add option to configure the "Files" section

rnhmjoj committed:
nixos/fontdir: add the directory to the xserver font paths

rnhmjoj commented:
Ok, I figured how to add the font path without manually doing `xset fp= `. I'll push the changes here.

hir12111 commented:
Nice, I'll test that in Xwayland, I think the behavior is a bit different. I suspect it doesn't handle compressed font files.

hir12111 commented:
Yes, as I said, Xwayland doesn't handle compressed font files (at least when they are specified via `xset`), the font is listed in a `XListFonts` call, but if you try to open the font, you'd get a null pointer.
The second issue I had is: I coundn't find a way to configure a FontPath in Xwayland, it seems like it doesn't consider `xorg.conf` FontPaths.
Probably they are not issues to fix in this PR, but I thought it was interesting to know.

rnhmjoj commented:
Unfortunately I don't use Xwayland so I can't really help.

hir12111 commented:
What do you think about uncompressing fonts in the X11-fonts scripts? it would solve an issue for Xwayland users

rnhmjoj commented:
Uhm, it's not very clean but if it's the only way I would add another option, off by default because it can be quite expensive if there are a lot of fonts installed.

If we are adding more options to this module we may also have to rename 
```
fonts.enableFontdir → fonts.fontDir.enable
fonts.fontDir.decompressFonts
```

EmilioPeJu committed:
nixos/fontdir: add group of options for fontDir

Renaming enableFontDir to fontDir.enable

EmilioPeJu committed:
nixos/fontdir: add option to decompress fonts

This will let Xwayland use the global font folder as font path

hir12111 commented:
Done, I tested the result in Xwayland and works fine.

rnhmjoj commented on nixos/modules/config/fonts/fontdir.nix:
```diff
@@ -10,6 +10,9 @@ let
     find ${toString config.fonts.fonts} -regex "$font_regexp" \
       -exec ln -sf -t "$out/share/X11/fonts" '{}' \;
     cd "$out/share/X11/fonts"
+    ${if config.fonts.fontDir.decompressFonts then ''
```
You could use [lib.optionalString](https://nixos.org/manual/nixpkgs/stable/#function-library.strings.optionalString) here.

hir12111 replied:
Fixed, thanks for letting me know about that function

hir12111 commented:
I found that it is possible to set a default font path in Xwayland by passing it in the configure phase.
Do you think it is worth doing it if this option is enable? the downside is Xwayland would be recompiled.

rnhmjoj commented:
It seems the font directory is currently unset (`--with-default-font-path=""`).
You make it default to something more sensible, but it shouldn't be NixOS-specific: Nixpkgs is intended to be used on other platforms as well.
As you said, it could also be overriden from the fontdir module but it would make the build of xwayland depend on `config.fonts.fonts`, which is not a good idea (rebuild every time you add/remove/update a font).

I would open an issue or ask [upstream](https://gitlab.freedesktop.org/xorg/xserver) why it's ignoring the `FontPath`.

hir12111 commented:
They are ignoring `FontPath` because their goal is to not need a configuration file, which is not a problem for most distributions because they put their fonts in one of the built-ins paths.
My proposal is to create a configuration parameter e.g: `config.xwayland.defaultFontPath` which would get set if `fontDir` is enabled.
This options would be passed to the xwayland derivation as an argument with a default value of empty string.
Do you agree with that approach?

rnhmjoj commented:
> They are ignoring FontPath because their goal is to not need a configuration file

So it's not reading /etc/X11/xorg.conf at all?

> This parameter would be passed to the xwayland derivation as an argument with a default value of empty string.

Works but would still happen by overriding the package using an overlay.  Being a configure flag it would force a rebuild every time. We could use `/run/current-system/sw/share/X11/fonts` to reduce the number of rebuilds to only one, though it won't be deterministic.

hir12111 commented:
> So it's not reading /etc/X11/xorg.conf at all?
That's right, Xwayland doesn't read it at all, I've confirmed that by running it with `strace` and not finding any `open` syscall to `xorg.conf`.

> Works but would still happen by overriding the package using an overlay. Being a configure flag it would force a rebuild every time. We could use `/run/current-system/sw/share/X11/fonts` to reduce the number of rebuilds to only one, though it won't be deterministic.

I can see that pointing to `/run/current-system` is done in a few derivations, is that considered a bad practice for this case?
I don't think the compilation process does anything with that path, and if it is passed as a string, I believe the build could still be deterministic.

I can give that approach a try, if it's not convincing enough in tests, we can simply remove the commit.

rnhmjoj commented:
Yes, it's done in a few places but should be done only for good reasons. Technically, it's an impurity because it makes (in this case xwayland) access a path that's outside the system closure. A notable example is `/run/opengl-driver` that is needed to not make every software depend on specific hardware/drivers of the running machine.

I don't have strong opinions; particularly in case I don't think any side effect will be so bad (not as in the opengl case, at least). 

> I can give that approach a try, if it's not convincing enough in tests, we can simply remove the commit.

Of course, feel free to experiment.

EmilioPeJu committed:
nixos/xwayland: add new module and allow configuring a default font path

- Add option `programs.xwayland.defaultFontPath`
- Modify sway to enable Xwayland

hir12111 commented:
I think it makes sense to have a new module section for `Xwayland` to hold the option `globalFontPath`  (and probably more options in future)
Enabling `fontDir` will enable this option by default, which in turn will enable `decompressFonts` to consider the issue I mentioned before.

rnhmjoj commented on nixos/modules/services/x11/xwayland.nix:
```diff
@@ -0,0 +1,43 @@
+{
+  options.services.xwayland = {
+
+    enable = mkEnableOption "Xwayland";
```
Add a (short) description of what it is/does. You can take a look at siminal options from `man configuration.nix`.

hir12111 replied:
Fixed

rnhmjoj commented on pkgs/servers/x11/xorg/xwayland.nix:
```diff
@@ -1,4 +1,4 @@
-{ stdenv, wayland, wayland-protocols, xorgserver, xkbcomp, xkeyboard_config, epoxy, libxslt, libunwind, makeWrapper, egl-wayland }:
+{ stdenv, wayland, wayland-protocols, xorgserver, xkbcomp, xkeyboard_config, epoxy, libxslt, libunwind, makeWrapper, egl-wayland, useGlobalFontPath ? false }:
```
Can you split this on multiple lines: it's better when looking at diffs. Also the options should be visually separate from input packages.

hir12111 replied:
Done

rnhmjoj commented on nixos/modules/services/x11/xwayland.nix:
```diff
@@ -0,0 +1,43 @@
+with lib;
+
+let
+  cfg = config.services.xwayland;
```
Since this is not strictly a service (in the running-as-a-daemon sense) it would more appropriate to put this option under the `programs` namespace and move the module to `nixos/modules/programs/`.

hir12111 replied:
Well, it depends on which definition of service we use:
- `Xwayland` could be run at boot time by the display manager
- `Xwayland` is running in the background, translating requests from X protocol to Wayland protocol (and vice-versa)
- `Xwayland` is actually `xorg-server` (which is considered as a service) compiled with specific configure options 

I've moved it to `programs`, however if you agree with the reasons I provided, I could move it back

hir12111 replied:
The file is still in `modules/services/x11`, but I thought that made sense because it's X11 stuff

rnhmjoj commented on nixos/modules/programs/sway.nix:
```diff
@@ -117,7 +125,9 @@ in {
     environment = {
-      systemPackages = [ swayPackage ] ++ cfg.extraPackages;
+      systemPackages = [ swayPackage ]
+        ++ (if cfg.useXwayland then [config.services.xwayland.package] else [])
```
Isn't the xwayland module already installing the package?

hir12111 replied:
whoops, sorry about that
Fixed!

rnhmjoj commented on pkgs/servers/x11/xorg/xwayland.nix:
```diff
@@ -1,4 +1,6 @@
-{ stdenv, wayland, wayland-protocols, xorgserver, xkbcomp, xkeyboard_config, epoxy, libxslt, libunwind, makeWrapper, egl-wayland }:
+{ stdenv, wayland, wayland-protocols, xorgserver, xkbcomp, xkeyboard_config
+, epoxy, libxslt, libunwind, makeWrapper, egl-wayland
+, useGlobalFontPath ? false }:
```
Why not make this a string directly? It would make changing the path for non-NixOS users easier.

hir12111 replied:
done, I renamed it to defaultFontPath

rnhmjoj commented on nixos/modules/services/x11/xwayland.nix:
```diff
@@ -0,0 +1,45 @@
+      Xwayland X server allows running X programs on a Wayland compositor.
+    '';
+
+    globalFontPath = mkOption {
```
This should also be a string, I think.

hir12111 replied:
Done and renamed to `defaultFontPath`

hir12111 commented:
Just let you know that in my last amendment there is no reference to `/run/current-system` anymore

rnhmjoj commented:
Can you move `nixos/modules/services/x11/xwayland.nix` to `nixos/modules/programs/`?

rnhmjoj commented on nixos/modules/services/x11/xwayland.nix:
```diff
@@ -0,0 +1,41 @@
+      type = types.str;
+      default = "";
+      description = ''
+        Default font path.
```
It may be worth mentioning that this causes the rebuild of xwayland.

hir12111 replied:
Done

rnhmjoj commented:
I still have to test the new module with sway but it looks pretty good.

As a last thing, when you're done, please take a look at the commit message [convention](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes) and reword the commits to make them consistent. I would also squash the first two commits.

hir12111 commented:
I did as suggested, and as a consequence of squashing the 2 first commits, your changes lost their signature, feel free to rebase if you want to add it again

rnhmjoj commented on nixos/modules/config/fonts/fontdir.nix:
```diff
@@ -21,27 +25,42 @@ in
-    };
+      decompressFonts = mkOption {
+        type = types.bool;
+        default = false;
```
```
    fonts.fontDir.decompressFonts = mkIf config.programs.xwayland.enable true;
```

This definition will easily cause a conflict. You could use `mkDefault` but It's better to simply do
```suggestion
        default = config.programs.xwayland.enable;
```

hir12111 replied:
done

rnhmjoj commented on nixos/modules/config/fonts/fontdir.nix:
```diff
@@ -21,27 +25,42 @@ in
+      FontPath "${x11Fonts}/share/X11/fonts"
+    '';
+
+    programs.xwayland.defaultFontPath = "${x11Fonts}/share/X11/fonts";
```
If xwayland is enabled setting `fonts.fontsDir.enable` will implicitly cause a rebuild and a user could be pretty clueless: I would move this to the xwayland module. It may be reasonable to set the default value of `programs.xwayland.defaultFontPath` to
```
optionalString config.fonts.fontDir.enable "/run/current-system/sw/share/X11/fonts"
```

hir12111 replied:
done

rnhmjoj commented on nixos/modules/programs/sway.nix:
```diff
@@ -104,6 +103,15 @@ in {
         Extra packages to be installed system wide.
       '';
     };
+
```
This options is pretty redundant/confusing. If you are trying to make a default option configurable you can simply set the value using `mkDefault`, see `programs.dconf.enable = mkDefault true;` just below.

hir12111 replied:
Done! Thanks for pointing that out

rnhmjoj reviewed (changes requested):
I tested sway and xwayland and found a couple of small issues...

hir12111 commented:
I rebased on top of master to fix one of the checks

rnhmjoj committed:
nixos/doc: document fonts.enableFontDir changes

rnhmjoj commented:
Thank you! I made some additional fixups:

- Added a `mkRenamedOptionModule` to make the option rename backward compatible
- Mentioned the changes in the release notes for 20.09
- Linked `/share/X11` when Xwayland is enabled.

rnhmjoj reviewed (approved):
I checked a final time with XWayland (in sway) and Xorg with xlsfonts: everything seems to work as expected. This could use another pair of eyes but it looks ready to go for me.

hir12111 commented:
Thank you for reviewing and for all the fixes.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/297

erikarvstedt commented on nixos/modules/programs/xwayland.nix:
```diff
@@ -0,0 +1,45 @@
+      default = optionalString config.fonts.fontDir.enable
+        "/run/current-system/sw/share/X11/fonts";
+      description = ''
+        Default font path.  Setting this option causes Xwayland to be rebuilt.
```
Nit:
```suggestion
        Default font path. Setting this option causes Xwayland to be rebuilt.
```

erikarvstedt commented on nixos/modules/services/x11/xserver.nix:
```diff
@@ -364,6 +365,13 @@ in
+        type = types.lines;
+        default = "";
+        example = ''FontPath "/path/to/my/fonts"'';
+        description = "Contents of the first Files section of the X server configuration file.";
```
```suggestion
        description = "Contents of the first <literal>Files</literal> section of the X server configuration file.";
```

hir12111 commented:
Hi @erikarvstedt and thanks for reviewing, I rebased one more time to have those changes in its related commit

rnhmjoj commented:
I think we are far too late for releasing this into 20.09. Can you move the release note into 21.03?

rnhmjoj commented:
Thank you for updating the note. The PR looked already good to me but I also run a bunch of X.org related tests to be use: everything seems ok. I'm going to merge.
